pkg/notifications/notifications.go

```diff
@@ -0,0 +1,447 @@
+	if err != nil {
+		return nil, err
+	}
+
+	ev, err := s.buildCommonEvent(res)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build event for %q,%q", res.HostName, res.ServiceName)
+	}
+
+	if h.PercentStateChangeEnd.Valid {
+		ev.Type = event.TypeFlappingEnd
+		ev.Message = fmt.Sprintf(
+			"Checkable stopped flapping (Current flapping value %.2f%% < low threshold %.2f%%)",
+			h.PercentStateChangeEnd.Float64, h.FlappingThresholdLow)
+	} else if h.PercentStateChangeStart.Valid {
+		ev.Type = event.TypeFlappingStart
+		ev.Message = fmt.Sprintf(
+			"Checkable started flapping (Current flapping value %.2f%% > high threshold %.2f%%)",
+			h.PercentStateChangeStart.Float64, h.FlappingThresholdHigh)
+		ev.Mute = types.MakeBool(true)
+		ev.MuteReason = "Checkable is flapping"
+	} else {
+		return nil, errors.New("flapping history entry has neither percent_state_change_start nor percent_state_change_end")
+	}
+
+	return ev, nil
+}
+
+// buildAcknowledgementHistoryEvent from an acknowledgment history entry.
+func (s *Source) buildAcknowledgementHistoryEvent(ctx context.Context, h *v1history.AcknowledgementHistory) (*event.Event, error) {
+	res, err := s.fetchHostServiceName(ctx, h.HostId, h.ServiceId)
+	if err != nil {
+		return nil, err
+	}
+
+	ev, err := s.buildCommonEvent(res)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build event for %q,%q", res.HostName, res.ServiceName)
+	}
+
+	if !h.ClearTime.Time().IsZero() {
+		ev.Type = event.TypeAcknowledgementCleared
+		ev.Message = "Checkable was cleared"
+
+		if h.ClearedBy.Valid {
+			ev.Username = h.ClearedBy.String
+		}
+	} else if !h.SetTime.Time().IsZero() {
+		ev.Type = event.TypeAcknowledgementSet
+
+		if h.Comment.Valid {
+			ev.Message = h.Comment.String
+		} else {
+			ev.Message = "Checkable was acknowledged"
+		}
+
+		if h.Author.Valid {
+			ev.Username = h.Author.String
+		}
+	} else {
+		return nil, errors.New("acknowledgment history entry has neither a set_time nor a clear_time")
+	}
+
+	return ev, nil
+}
+
+// worker is the background worker launched by NewNotificationsSource.
+func (s *Source) worker() {
+	defer s.ctxCancel()
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+
+		case entity, more := <-s.inputCh:
+			if !more { // Should never happen, but just in case.
+				s.logger.Debug("Input channel closed, stopping worker")
+				return
+			}
+
+			var ev *event.Event
+			var eventErr error
+
+			// Keep the type switch in sync with syncPipelines from pkg/icingadb/history/sync.go
+			switch h := entity.(type) {
+			case *v1history.NotificationHistory:
+				// Ignore for the moment.
+				continue
+
+			case *v1history.StateHistory:
+				if h.StateType != common.HardState {
+					continue
+				}
+
+				ev, eventErr = s.buildStateHistoryEvent(s.ctx, h)
+
+			case *v1history.DowntimeHistory:
+				ev, eventErr = s.buildDowntimeHistoryEvent(s.ctx, h)
+
+			case *v1history.CommentHistory:
+				// Ignore for the moment.
+				continue
```

Comment:
The comment (as in code comment here) would be more helpful if it said when this should be implemented / what is currently missing to implement it. And it should probably include the TODO keyword so that someone might actually find it later.